### ISSUE TITLE
SHA1_init missing when using PyInstaller on OSX

### ISSUE THREAD

**Issue by Robby0318** [pyinstaller/pyinstaller#5583]
## Description of the issue
There is no problem running the source code directly with python3 under OSX 11.1
When running the executable generated by PyInstaller, I get the following error message. 

 ```
File "gpsoauth/__init__.py", line 97, in perform_master_login
File "gpsoauth/google.py", line 51, in signature
File "Cryptodome/Cipher/PKCS1_OAEP.py", line 117, in encrypt
lHash = self._hashObj.new(self._label).digest()
 File "Cryptodome/Hash/SHA1.py", line 158, in new
 return SHA1Hash().new(data)
 File "Cryptodome/Hash/SHA1.py", line 74, in __init__
 result = _raw_sha1_lib.SHA1_init(state.address_of())
 File "ctypes/__init__.py", line 369, in __getattr__
File "ctypes/__init__.py", line 374, in __getitem__
AttributeError: dlsym(0x7fc07e7ee6d0, SHA1_init): symbol not found
 ```

After testing, it is found that the `sha1_init` symbol is in the source code `Cryptodome/Hash/_SHA1.abi3.so` library, and the `_sha1.cpython-39m-darwin.so` generated by Pyinstaller does not have the `sha1_init` symbol.

 ## Context information (for bug reports)
* pyinstaller 5.0dev
* python 3.9.1
* OSX 11.1
* There is no problem with the executable file generated by pyinstaller under the win platform

**Comment by bwoodsend** [comment]
Can you please fill out the [issue template](https://github.com/pyinstaller/pyinstaller/blob/develop/.github/ISSUE_TEMPLATE/bug_report.md).

**Comment by Robby0318** [comment]
> Can you please fill out the [issue template](https://github.com/pyinstaller/pyinstaller/blob/develop/.github/ISSUE_TEMPLATE/bug_report.md).

ok, it has been changed

**Comment by bwoodsend** [comment]
Can you try using `collect_dynamic_libs()`? Edit you `.spec` file to put:
```python
from PyInstaller.utils.hooks import collect_dynamic_libs
```
at the top and change `binaries=[]` to `binaries=collect_dynamic_libs("Cryptodome")`. Then rebuild using:
```
pyinstaller your-project-name.spec
```
That at least rules ensures the right libraries are collected.

**Comment by rokm** [comment]
Uf, this is a nasty subtle bug... 

There is a `Cryptodome/Hash/_SHA1.abi3.so˙ library that's properly collected and which is supposed to be loaded by `pycryptodomex`, but we end up with handle to `_sha1.cpython-38-darwin.so` that's collected in the root dist folder (and comes from python itself). 

The loading is done via `cffi` or, if that's not available, via `ctypes` as a fallback.

In both cases, the library is loaded by calling [utility function](https://github.com/Legrandin/pycryptodome/blob/7bc6d4d548596e5bfa3da0dd253a068b403b7c51/lib/Crypto/Hash/SHA1.py#L29) with fully-qualified name, e.g., `Crypto.Hash._SHA1`. This `load_pycryptodome_raw_lib` helper [decomposes the fully-qualified name into directory components and filename, and then tries to find the library using different extensions](https://github.com/Legrandin/pycryptodome/blob/7bc6d4d548596e5bfa3da0dd253a068b403b7c51/lib/Crypto/Util/_raw_api.py#L281-L300).

The extension suffices that are tested come from `importlib.machinery.EXTENSION_SUFFIXES`, and are:
```
['.cpython-38-darwin.so', '.abi3.so', '.so']
```
This means than unfrozen version first tries to load
```
/Users/[...]/venv/lib/python3.8/site-packages/Cryptodome/Util/../Hash/_SHA1.cpython-38-darwin.so
```
which fails as it does not exist. Then it tries
```
/Users/[...]/venv/lib/python3.8/site-packages/Cryptodome/Util/../Hash/_SHA1.abi3.so
```
which succeeds.

In frozen version, however, the first call with `_MEIPASS/Cryptodome/Util/../Hash/_SHA1.cpython-38-darwin.so` actually succeeds, but it returns handle to `_MEIPASS/_SHA1.cpython-38-darwin.so`.

In `ctypes` codepath, this happens because our loader explicitly checks if the library with requested basename exists in _MEIPASS and if it does, returns it. I.e.,

https://github.com/pyinstaller/pyinstaller/blob/93285ece5a02932c6dac8f018bf107e7618d7d3c/PyInstaller/loader/pyiboot01_bootstrap.py#L126-L131

called from

https://github.com/pyinstaller/pyinstaller/blob/93285ece5a02932c6dac8f018bf107e7618d7d3c/PyInstaller/loader/pyiboot01_bootstrap.py#L140-L146

The same thing happens with `cffi`, although the mechanism is different there, but I suspect it involves this:

https://github.com/pyinstaller/pyinstaller/blob/93285ece5a02932c6dac8f018bf107e7618d7d3c/PyInstaller/loader/pyiboot01_bootstrap.py#L193-L199

So long story short, we find a library that should not be found. (Although on the other hand, I also understand why also checking _MEIPASS might be needed in other scenarios).

**Comment by rokm** [comment]
@Robby0318, since your log indicates that you're using `ctypes` codepath, as a work-around you can try disabling that `_frozen_name` rewrite in your local PyInstaller installation (in `site-packages/PyInstaller/loader/pyiboot01_bootstrap.py`). 

For example, try changing:
```python
def _frozen_name(name): 
    if name: 
        [...]
```
to 
```python
def _frozen_name(name): 
    if name and not os.path.isabs(name):
        [...]
```
and rebuild your application. (Although whether this will work depends a lot on behavior of other packages that you're using...)

**Comment by tom-james-watson** [comment]
I can confirm that my issue (https://github.com/pyinstaller/pyinstaller-hooks-contrib/issues/109) was fixed by https://github.com/pyinstaller/pyinstaller/pull/5604. Thanks, @rokm!
